timesync/peersync/sync.go

```diff
@@ -0,0 +1,330 @@
+	defer s.log.With().Debug("exiting sync background worker")
+	for {
+		prs, err := s.peersWatcher.waitPeers(s.ctx, s.config.RequiredResponses)
+		if err != nil {
+			return err
+		}
+		s.log.With().Info("starting time sync round with peers",
+			log.Uint64("round", round),
+			log.Int("peers_count", len(prs)),
+			log.Uint32("errors_count", atomic.LoadUint32(&s.errCnt)),
+		)
+		rctx, cancel := context.WithTimeout(s.ctx, s.config.RoundTimeout)
+		offset, err := s.GetOffset(rctx, round, prs)
+		cancel()
+
+		var timeout time.Duration
+		if err == nil {
+			if offset > s.config.MaxClockOffset || (offset < 0 && -offset > s.config.MaxClockOffset) {
+				s.log.With().Error("peers offset is larger than max allowed clock difference",
+					log.Uint64("round", round),
+					log.Duration("offset", offset),
+					log.Duration("max_offset", s.config.MaxClockOffset),
+				)
+				if atomic.AddUint32(&s.errCnt, 1) == uint32(s.config.MaxOffsetErrors) {
+					return ErrPeersNotSynced
+				}
+			} else {
+				s.log.With().Info("peers offset is within max allowed clock difference",
+					log.Uint64("round", round),
+					log.Duration("offset", offset),
+					log.Duration("max_offset", s.config.MaxClockOffset),
+				)
+				atomic.StoreUint32(&s.errCnt, 0)
+			}
+			timeout = s.config.RoundInterval
+		} else {
+			s.log.With().Error("failed to fetch offset from peers", log.Err(err))
+			timeout = s.config.RoundRetryInterval
+		}
+
+		round++
+		if timer == nil {
+			timer = time.NewTimer(timeout)
+		} else {
+			timer.Reset(timeout)
+		}
+		select {
+		case <-s.ctx.Done():
+			return s.ctx.Err()
+		case <-timer.C:
+		}
+	}
+}
+
+// GetOffset computes offset from received response. The method is stateless and safe to use concurrently.
+func (s *Sync) GetOffset(ctx context.Context, id uint64, prs []p2pcrypto.PublicKey) (time.Duration, error) {
+	var (
+		responses = make(chan Response, len(prs))
+		round     = round{
+			ID:                id,
+			Timestamp:         s.time.Now().UnixNano(),
+			RequiredResponses: s.config.RequiredResponses,
+		}
+		request = Request{ID: id}
+	)
+	buf, err := types.InterfaceToBytes(&request)
+	if err != nil {
+		s.log.With().Panic("can't encode request to bytes", log.Err(err))
+	}
+	var (
+		errCnt    int
+		respCount int
+	)
+	for _, peer := range prs {
+		if err := s.srv.SendRequest(ctx, server.RequestTimeSync, buf, peer, func(buf []byte) {
+			var response Response
+			if err := types.BytesToInterface(buf, &response); err != nil {
+				s.log.Debug("can't decode response", log.Binary("response", buf), log.Err(err))
+				return
+			}
+			select {
+			case <-ctx.Done():
+			case responses <- response:
+			}
+		}, func(error) {}); err != nil {
+			s.log.Debug("can't send request to peer", log.Err(err))
+			errCnt++
+		}
+		if d := len(prs) - errCnt; d < s.config.RequiredResponses || d == 0 {
+			return 0, fmt.Errorf("%w: failed to send requests to %d peers from %d", ErrTimesyncFailed, errCnt, len(prs))
+		}
+	}
+
+	wait := func() {
```

Comment:
if `len(prs)-errCnt` < `s.config.RequiredResponses` we don't even have to wait for responses

Reply:
we are not. i return on L306